ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,254 @@
+    => Show (Message (VersionNegotiationProtocol vNumber vBlob) from to)
+
+-- |
+-- Version negotiation failures
+data VersionNegotiationError =
+    -- |
+    -- Remote peer rejected version
+      VersionNegotiationRejected
+
+    -- |
+    -- Protocol error, remote peer accepted a version but send back unknown version.
+    --
+    | VersionNegotiationUnknownVersion
+
+    -- | Error while decoding version data on agreed version.
+    --
+    | VersionNegotiationDecodeError
+
+  deriving (Eq, Ord, Show)
+
+encodeVersions
+  :: forall vNumber extra r vBlob.
+     (forall vData. extra vData -> vData -> vBlob)
+  -> Versions vNumber extra r
+  -> Map vNumber vBlob
+encodeVersions encoder (Versions vs) = go <$> vs
+    where
+      go :: Sigma (Version extra r) -> vBlob
+      go (Sigma vData Version {versionExtra}) = encoder versionExtra vData
+
+-- |
+-- Version negotiation client which offers @'KnownVersions' vNumber r@ to the
+-- remote peer.
+--
+-- TODO: GADT encoding of the client (@VersionNegotiation.Client@ module).
+versionNegotiationClientPeer
+  :: Ord vNumber
+  => (forall vData. extra vData -> vData -> CBOR.Term)
+  -> (forall vData. extra vData -> CBOR.Term -> Maybe vData)
+  -> Versions vNumber extra r
+  -> Peer (VersionNegotiationProtocol vNumber CBOR.Term) AsClient StPropose m (Either VersionNegotiationError r)
+versionNegotiationClientPeer encodeData decodeData versions =
+  -- send known versions
+  Yield (ClientAgency TokPropose) (MsgProposeVersions $ encodeVersions encodeData versions) $
+
+    Await (ServerAgency TokConfirm) $ \msg -> case msg of
+
+      -- the server refused common highest version
+      MsgRefuse ->
+        Done TokDone (Left VersionNegotiationRejected)
+
+      -- the server didn't recognised the common highest version; either the
+      -- intersection was empty all the decoding failed.
+      MsgUnknownVersion ->
+        Done TokDone (Left VersionNegotiationUnknownVersion)
+
+      -- the server accepted a version, sent back the version number and its
+      -- version data blob
+      MsgAcceptVersion vNumber vBlob ->
+        case vNumber `Map.lookup` getVersions versions of
+          Nothing -> Done TokDone (Left VersionNegotiationUnknownVersion)
+          Just (Sigma vData version) ->
+            case decodeData (versionExtra version) vBlob of
+
+              Nothing ->
+                Done TokDone (Left VersionNegotiationDecodeError)
+
+              Just vData' ->
+                Done TokDone $ Right $ runApplication (versionApplication version) vData vData'
+
+-- |
+-- Version negotiation server which accepts highest version offered by the more
+-- peer that also belongs to @knownVersion@.
+--
+-- TODO: GADT encoding of the server (@VersionNegotiation.Server@ module).
+versionNegotiationServerPeer
+  :: Ord vNumber
+  => (forall vData. extra vData -> vData -> vBlob)
+  -> (forall vData. extra vData -> vBlob -> Maybe vData)
+  -> (forall vData. extra vData -> vData -> vData -> Bool)
```

Comment:
Curious about whether this is needed. Is it not OK to always accept when the version numbers match and the blob decodes successfully?

Reply:
I think it is not ok: primary example is trying to spent test-net ada on a main-net ;).

Reply:
^ that would be impossible by the consensus validation, but we should not allow connecting across networks with different network magic.

Reply:
Yeah I was going to say, surely the protocol version is not the principal guard against such a bug...

I would have thought that the _protocol version_ is the same for testnet or for mainnet. Which *net a thing is running is a higher-level concept than which *network protocol* it's running, no?

Reply:
In other words, I'm of the opinion that the testnet/mainnet/whatevernet distinction should *not* be made at the network protocol level.

Reply:
And I don't believe my opinion is controversial: two nodes can run the very same network protocol, yet track different *nets. Is there a very good reason to conflate these two things?

Reply:
We're not currently planning to have nodes connecting to both testnet and mainnet at once, and we do want to avoid nodes accidentally connecting to the wrong ones.

I also don't consider it part of the protocol version as such, but it's quite reasonable I think to see it as a paramater. So I think it's fine to have the "magic" be a paramater used in the version negoritation and the ability to reject on the basis of mismatch.

Reply:
I added `RefuseReason` which is used to reply / return from the client/server.  Also the `accept` callback now is supposed to return an `Accept` type which is isomorphic to `Maybe Text`, giving a reason why the server refused to run a given version.

Reply:
> So I think it's fine to have the "magic" be a paramater used in the version negoritation and the ability to reject on the basis of mismatch.

Sure, it's fine, it just makes less sense than not doing it. It introduces a case in which the peers actually *do* speak the same network protocol, yet they refuse based on a mismatched version.